### PR TITLE
Add FSQ implementation

### PR DESCRIPTION
**Pull request by sekstini**
TODO:
- [x] Verify correctness
- [x] Add usage example

Notes:
* Torch doesn't support uint32 yet, so we use int32. Should be fine.
* Part of the offset calculation (in the `bound` function) is missing. Just took copilot's suggestion for now.
* Fixed some grammatical errors (indexes -> indices, and incorrect docstring)

### CODE REVIEW

**Comment by lucidrains** [comment]
nice! give it a test drive with the cifar script in the examples folder

@kashif  ^

**Comment by sekstini** [comment]
Still not 100% sure this is correct, but initial results seem promising.

```bash
 vq :: rec loss: 0.114 | cmt loss: 0.001 | active %: 21.094

fsq :: rec loss: 0.111 | active %: 58.333
```

Not parameter matched, so losses aren't really representative, but interesting to see the higher codebook usage.

**Comment by lucidrains** [comment]
do you have the training curves for each?

**Comment by sekstini** [comment]
losses
active_codes

**Comment by sekstini** [comment]
Probably isn't representative of FSQ performance in general, but it does seem to be working ^^

**Comment by sekstini** [comment]
Ok wait, strike that. Reading the paper a bit more closely, the offsets should be flipped relative to what I put there initially. Doing so nets a much higher active %.

File: vector_quantize_pytorch/finite_scalar_quantization.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        self.register_buffer("_basis", _basis)
+
+        codebook_size = self._levels.prod()
+        implicit_codebook = self.indices_to_codes(torch.arange(codebook_size))
+        self.register_buffer("implicit_codebook", implicit_codebook)
+
+    def forward(self, z: torch.Tensor) -> torch.Tensor:
+        zhat = self.quantize(z)
+        indices = self.codes_to_indices(zhat)
+        return self.implicit_codebook[indices], indices
```
Comment (lucidrains) [inline vector_quantize_pytorch/finite_scalar_quantization.py:32]: shouldn't you return zhat here? so that the gradients actually straight through back?

Comment (sekstini) [reply vector_quantize_pytorch/finite_scalar_quantization.py:32]: Yeah, probably. The paper didn't have a forward function, and I wasn't sure what the expected behaviour should be.

Comment (lucidrains) [reply vector_quantize_pytorch/finite_scalar_quantization.py]: try it with `return zhat, indices` and rerun the cifar experiment

File: vector_quantize_pytorch/finite_scalar_quantization.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    def forward(self, z: torch.Tensor) -> torch.Tensor:
+        zhat = self.quantize(z)
+        indices = self.codes_to_indices(zhat)
+        return zhat, indices
+
+    def bound(self, z: torch.Tensor, eps: float = 1e-3) -> torch.Tensor:
+        """Bound `z`, an array of shape (..., d)."""
+        half_l = (self._levels - 1) * (1 - eps) / 2
+        offset = torch.where(self._levels % 2 == 0, 0.5, 0.0)
```
Comment (sekstini) [inline vector_quantize_pytorch/finite_scalar_quantization.py:37]: Should probably message the authors and get a confirmation that this is correct?

File: vector_quantize_pytorch/finite_scalar_quantization.py (hunk 1 above)
Comment (sekstini) [reply vector_quantize_pytorch/finite_scalar_quantization.py]: Seems I forgot to mention it, but I couldn't find the cifar example. I'm testing on the modified autoencoder example that's included in this PR.

Comment (lucidrains) [reply vector_quantize_pytorch/finite_scalar_quantization.py]: oh yup, that's the one I meant 😂 fashion mnist, cifar, similar toytasks

File: vector_quantize_pytorch/finite_scalar_quantization.py (hunk 2 above)
Comment (lucidrains) [reply vector_quantize_pytorch/finite_scalar_quantization.py:37]: yea it looks like their code sample is marred. the authors will probably have to fix that before submitting to a conference

Comment (lucidrains) [reply vector_quantize_pytorch/finite_scalar_quantization.py:37]: you can def email them to let them know about it

File: examples/autoencoder_fsq.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                nn.Upsample(scale_factor=2, mode="nearest"),
+                nn.Conv2d(16, 1, kernel_size=3, stride=1, padding=2),
+            ]
+        )
+        return
+
+    def forward(self, x):
+        for layer in self.layers:
+            if isinstance(layer, FSQ):
+                x_flat, indices = layer(x)
```
Comment (sekstini) [inline examples/autoencoder_fsq.py:46]: Should `x_flat` just be `x` here? Otherwise the changes to the forward won't matter right?

Comment (lucidrains) [reply examples/autoencoder_fsq.py:46]: yup, that should just be `x`

**Comment by sekstini** [comment]
plot

Okay, this looks fine I think.

**Comment by lucidrains** [comment]
@sekstini getting there! we can let @kashif do a review before getting it merged! thanks for beasting through this

File: vector_quantize_pytorch/finite_scalar_quantization.py (hunk 2 above)
Comment (fab-jul) [reply vector_quantize_pytorch/finite_scalar_quantization.py:37]: Thanks :) This is the correct completion.

File: vector_quantize_pytorch/finite_scalar_quantization.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import torch.nn as nn
+
+
+def round_ste(z: torch.Tensor) -> torch.Tensor:
+    """Round with straight through gradients."""
+    zhat = z.round()
+    return z + (zhat - z).detach()
+
+
+class FSQ(nn.Module):
```
Comment (fab-jul) [inline vector_quantize_pytorch/finite_scalar_quantization.py:16]: Thanks so much for porting this! Do you mind if we link this repo in the next version and our own public code release?

Comment (fab-jul) [reply vector_quantize_pytorch/finite_scalar_quantization.py:16]: LMK if you are also planning to update the README and I can send some figs.

Comment (kashif) [reply vector_quantize_pytorch/finite_scalar_quantization.py:16]: please do! and i believe @lucidrains and @sekstini will appreciate it

**Comment by fab-jul** [comment]
> plot
> 
> Okay, this seems fine for an example. Curious to see if it works in more realistic scenarios too.

Cool to see the high util out of the box :)

File: vector_quantize_pytorch/finite_scalar_quantization.py (hunk 4 above)
Comment (sekstini) [reply vector_quantize_pytorch/finite_scalar_quantization.py:16]: Yup, go ahead 👍

**Comment by kashif** [comment]
@sekstini can you also kindly add a section in the README and the appropriate bibtex entry as #73 had started?

**Comment by sekstini** [comment]
> > plot
> > Okay, this seems fine for an example. Curious to see if it works in more realistic scenarios too.
> 
> Cool to see the high util out of the box :) I assume this is a fairly shallow AE? Probably the loss gap could be a bit smaller but looks like it's WAI! nice

Yeah, the test models are both shallow, and tiny at ~10k parameters. Don't expect this to realistically reflect performance at all.



> @sekstini can you also kindly add a section in the README and the appropriate bibtex entry as #73 had started?

Will do 👍

File: vector_quantize_pytorch/finite_scalar_quantization.py (hunk 4 above)
Comment (sekstini) [reply vector_quantize_pytorch/finite_scalar_quantization.py:16]: @fab-jul

> LMK if you are also planning to update the README and I can send some figs.

That would be great 🙏

**Comment by fab-jul** [comment]
> Yeah, the test models are both shallow, and tiny at ~10k parameters. Don't expect this to realistically reflect performance at all. [This is the FSQ network](https://github.com/lucidrains/vector-quantize-pytorch/pull/74/files#diff-3454bc5819476913301195605ee1fea904ad3eb684bec694c7faefb436224138R28-R38).

Makes sense! Yeah one thing with FSQ is that because your codebook is fixed, you need some decent number of layers before and after the quantizer. Usually this is the case (eg the VQ-GANs people train are fairly deep).

Anyway, I think the results you see are what we expect! Good stuff.

For the figures, I revisited your README and it seems it is usually mostly the code, so maybe simply using the cube ([PDF here](https://github.com/lucidrains/vector-quantize-pytorch/files/12762443/overview_v2_a.pdf)) would be enough or already too much. Your call :)

In the README of our upcoming mini-repo, I also added the table, but again, this might be out of place in your README:

|                  | VQ | FSQ |
|------------------|----|-----|
| Quantization     | argmin_c \|\| z-c \|\| | round(f(z)) |
| Gradients        | Straight Through Estimation (STE) | STE |
| Auxiliary Losses | Commitment, codebook, entropy loss, ... | N/A |
| Tricks           | EMA on codebook, codebook splitting, projections, ...| N/A |
| Parameters       | Codebook | N/A |

source:

```
|                  | VQ | FSQ |
|------------------|----|-----|
| Quantization     | argmin_c \|\| z-c \|\| | round(f(z)) |
| Gradients        | Straight Through Estimation (STE) | STE |
| Auxiliary Losses | Commitment, codebook, entropy loss, ... | N/A |
| Tricks           | EMA on codebook, codebook splitting, projections, ...| N/A |
| Parameters       | Codebook | N/A |
```

**Comment by sekstini** [comment]
I added the figure, but not sure if we want the table as well.

Could you take a look and perhaps write some text for the entry @lucidrains?

**Comment by lucidrains** [comment]
@sekstini yea, let's add the table! defer to Fabian of course but i would say

```
This work out of Google Deepmind aims to vastly simplify the way vector quantization is done for generative modeling, removing the need for commitment losses, EMA updating of the codebook, as well as tackle the issues with codebook collapse or insufficient utilization. They simply round each scalar into discrete levels with straight through gradients; the codes become uniform points in a hypercube.
```

**Comment by lucidrains** [comment]
lgtm! releasing

thank you @sekstini !

**Comment by sekstini** [comment]
@lucidrains Oh, apparently I pointed this as the fsq branch, so you might need to merge it into master

**Comment by lucidrains** [comment]
@sekstini yup, no problem, thank you! 🙏

**Comment by fab-jul** [comment]
Thanks everyone! I'll add a link to this repo in our next revision, and from our own (upcoming) public code.

Gonna use this link: https://github.com/lucidrains/vector-quantize-pytorch#finite-scalar-quantization

**Comment by dribnet** [comment]
Appreciate this quick port and example code. Just thought I would add that the reconstructions look from the FashionMNSIT example seem reasonable 

figgy1

However a bit of surprise noticing it fails if switching back to MNIST

figgy2

Changing levels/seeds didn't help, so perhaps as @fab-jul mentioned it's just a case of needing more layers before/after quantizer, etc. for some cases.

**Comment by sekstini** [comment]
@dribnet Interesting. Not sure why we would see this particular failure mode here, but I made an example of residual vector quantization where it's working: https://gist.github.com/sekstini/7f089f71d4b975ec8bde37d878b514d0.

**Comment by lucidrains** [comment]
@dribnet adding more layers seems testable

**Comment by lucidrains** [comment]
LFQ https://arxiv.org/abs/2310.05737 looks simliar?

**Comment by lucidrains** [comment]
nevermind, it is slightly different, will be adding

**Comment by fab-jul** [comment]
> nevermind, it is slightly different, will be adding

It‘s FSQ with levels = 2 plus an entropy maximization loss, IIUC.

**Comment by lucidrains** [comment]
yes, FSQ generalizes it, save for the entropy loss. no ablation of the entropy loss, so unsure how necessary it is, but i'll add it. perhaps that will be their contribution

**Comment by lucidrains** [comment]
@fab-jul congrats either way! this could be a winner

**Comment by fab-jul** [comment]
@lucidrains thanks!

for us, levels = 2 was suboptimal but I can see how an entropy maximization would fix that. I wonder if that also improves FSQ (although ofc the goal of our paper was to get rid of all aux losses haha)

**Comment by lucidrains** [comment]
> @dribnet Interesting. Not sure why we would see this particular failure mode here, but I made a toy example of residual vector quantization where it's working: https://gist.github.com/sekstini/7f089f71d4b975ec8bde37d878b514d0.
> 
> residual_fsq

you should pair up with an audio researcher and think hard about this (residual fsq), maybe do a tweetstorm or blogpost if you see anything but do not have bandwidth to write a paper. i'm sure many will be thinking along these lines, looking at the new paper

**Comment by sekstini** [comment]
> @sekstini you should pair up with an audio researcher and think hard about this (residual fsq), maybe do a tweetstorm or blogpost if you see anything but do not have bandwidth to write a paper. i'm sure many will be thinking along these lines, looking at the new paper

Funny you mention it, as I'm experimenting with a vocoder based on this idea in this very moment ^^
I should definitely improve my tweeting game though.

**Comment by lucidrains** [comment]
@sekstini are you seeing good results?

thinking about a soundstream variation with multi-headed LFQ

**Comment by sekstini** [comment]
@lucidrains

> @sekstini are you seeing good results?

Not really, but definitely not because of FSQ.

I have exclusively been toying with various parallel encoding schemes, which I'm guessing are difficult for the model to learn, and I suspect residual quantization would work a lot better.

**Comment by lucidrains** [comment]
@sekstini ah got it, thanks!

**Comment by lucidrains** [comment]
@sekstini yea, if LFQ pans out over at [magvit2](https://github.com/lucidrains/magvit2-pytorch), i'll do some improvisation here and maybe someone can do the hard experimental work.

**Comment by lucidrains** [comment]
@sekstini almost done! https://github.com/lucidrains/vector-quantize-pytorch/pull/80

**Comment by sekstini** [comment]
> @sekstini almost done! #80

Neat! I made some decent progress on my vocoder with "parallel FSQ", but I'd be interested in swapping this in to compare performance. Feel free to tag me when it's done.

**Comment by lucidrains** [comment]
ok it is done, integrated in soundstream over [here](https://github.com/lucidrains/audiolm-pytorch/commit/c748fcdb565964bc562277bd73fbeb2e5df0ffca)

**Comment by lucidrains** [comment]
may be of interest! https://github.com/lucidrains/magvit2-pytorch/issues/4

**Comment by sekstini** [comment]
@mueller-franzes You may find [this comment](https://github.com/lucidrains/vector-quantize-pytorch/issues/94#issuecomment-1858833682) by Fabian interesting.

> Should it be (1 + eps) instead of (1-eps)?

Makes sense to me. I copied it directly from the paper, but at the time there was a `tan` instead of `atanh` there, so I didn't notice any issues while testing levels = 2.

As a side note, you may want to check out LFQ if you're interested in the levels = 2 case in particular.

**Comment by lucidrains** [comment]
@mueller-franzes @sekstini maybe we should just enforce odd levels for now for FSQ until a correction to the paper comes out?

**Comment by sekstini** [comment]
> @mueller-franzes @sekstini maybe we should just enforce odd levels for now for FSQ until a correction to the paper comes out?

Other than the asymmetry being weird, levels > 2 seems fine (actually most of my code has been using even values).

I think switching to (1 + eps) or enforcing levels > 2 makes sense.

**Comment by mueller-franzes** [comment]
Thank you both for the super quick response! That's a good tip, I'll have a look at LFQ next.

**Comment by lucidrains** [comment]
@sekstini oh you are right, it is a levels == 2 problem

ok let's just go with `1 + eps`! thank you both
